### PR TITLE
Support storage classes on Kubernetes infra.

### LINKED ISSUE #12816: Feature request: storage class support for Kubernetes infra.
### Description

Some infra admins may want to assign the specific storage class to Che.
One of the major case is: Their K8s cluster has already running for another services and the default storage class is not match to Che workspaces.

**OS and version:**    

CHE-7.0.0-beta-2.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by monaka**
### What does this PR do?

Supports storage classes on Kubernetes infra.

### What issues does this PR fix or reference?

Will fix #12816 

#### Changelog 

Supports storage classes on Kubernetes infra.

#### Docs PR

To be done. (this is the reason why this PR is under draft.)

### CODE REVIEW

**Comment by monaka** [comment]
ci-build

**Comment by monaka** [comment]
It seems there is no description about properties for PVCs in the repistory eclipse/che-docs.
Where should I write to? (or, no need?)

**Comment by sleshchenko** [comment]
@monaka Please take a look che-docs repository (for example this page https://github.com/eclipse/che-docs/blob/master/src/main/pages/che-6/setup-kubernetes/kubernetes-or-openshift-admin-guide.adoc#how-the-che-server-uses-pvcs-and-pvs-for-storage ). I think the information about your feature would be useful there.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java [hunk 1]
```diff
@@ -95,12 +96,26 @@ public static void putSelector(Service target, String key, String value) {
...
-        .withNewSpec()
-        .withAccessModes(accessMode)
+    return newPVC(name, accessMode, quantity, null);
+  }
+
+  /**
+   * Returns new instance of {@link PersistentVolumeClaim} with specified name, accessMode, quantity
+   * and storageClassName.
+   */
+  public static PersistentVolumeClaim newPVC(
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java:122]: What do you think about rewriting this method in the following way?
```java
  public static PersistentVolumeClaim newPVC(
      String name, String accessMode, String quantity, String storageClassName) {
    return new PersistentVolumeClaimBuilder()
        .withNewMetadata()
        .withName(name)
        .endMetadata()
        .withNewSpec()
        .withAccessModes(accessMode)
        .withStorageClassName(isNullOrEmpty(storageClassName) ? null : storageClassName)
        .withNewResources()
        .withRequests(ImmutableMap.of(STORAGE_PARAM, new Quantity(quantity)))
        .endResources()
        .endSpec()
        .build();
  }
```

**Comment by sleshchenko** [comment]
:+1: Thanks for the contribution. Good job. Please make your that you made changes to all PVC strategies, looks like changes for UniquePVCStrategy is missing https://github.com/eclipse/che/pull/12818#discussion_r261938741

**Comment by monaka** [comment]
@sleshchenko YesI read the doc you pointed in [your comment](https://github.com/eclipse/che/pull/12818#issuecomment-469138953) before. There describes about strategies level only. I think the storage class is deeper level than there... (and it's a little bit hard to write all undocumented properties ;-)

**Review by garagatyi — COMMENTED** [review]
Overall looks good. Please, address @sleshchenko comments

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategyTest.java [hunk 2]
```diff
@@ -113,6 +115,7 @@ public void shouldReturnPVCPerWorkspace() throws Exception {
     // then
     assertEquals(commonPVC.getMetadata().getName(), PVC_NAME_PREFIX + "-" + WORKSPACE_ID);
     assertEquals(commonPVC.getMetadata().getLabels().get(CHE_WORKSPACE_ID_LABEL), WORKSPACE_ID);
+    assertEquals(commonPVC.getSpec().getStorageClassName(), PVC_STORAGE_CLASS_NAME);
```
Comment (amisevsk) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategyTest.java:119]: Consider adding test cases to assert the default behavior here as well.

**Review by amisevsk — COMMENTED** [review]
Looks good, but as others have mentioned it needs to be applied to the unique strategy as well.

One question: what happens if storage class property is invalid / not supported? Is the error returned readable / sensible?

**Comment by monaka** [comment]
@amisevsk 

> One question: what happens if storage class property is invalid / not supported? Is the error returned readable / sensible?

AFAIK, recent K8s controllers will always support the storage class itself.

There may be several `invalid` cases. The most popular case will be a misspell (like 'defaulto' instead of 'default').
In this case, the controller will keep the status of the PVC `pending`. So all deployments will be timeout.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 3]
```diff
@@ -430,6 +430,10 @@ che.infra.kubernetes.pvc.precreate_subpaths=true
 # Defines the name of Persistent Volume Claim for che workspaces.
 che.infra.kubernetes.pvc.name=claim-che-workspace
 
+# Defines the storage class of Persistent Volume Clami for the workspaces.
```
[the remark this answers is not in the document]
Comment (monaka) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Thanks.I fixed it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 4]
```diff
@@ -129,7 +132,7 @@ public CommonPVCStrategy(
    * @return pvc that should be used for the specified runtime identity
    */
   protected PersistentVolumeClaim createCommonPVC(String workspaceId) {
-    return newPVC(configuredPVCName, pvcAccessMode, pvcQuantity);
+    return newPVC(configuredPVCName, pvcAccessMode, pvcQuantity, pvcStorageClassName);
```
[the remark this answers is not in the document]
Comment (monaka) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:135]: I see. I added some code to `PVCProvisioner.java` and fixed for passing tests.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java (hunk 1 above)
Comment (monaka) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java:122]: I've not tested but I think your suggestion won't work fine.

Referring to https://kubernetes.io/blog/2017/03/dynamic-provisioning-and-storage-classes-kubernetes/, 

> * If storageClassName is not specified in the PVC, the default storage class will be used for provisioning.
>    * Existing, “Available”, PVs that do not have the default storage class label will not be considered for binding to the PVC
> * If storageClassName is set to an empty string (‘’) in the PVC, no storage class will be used (i.e.; dynamic provisioning is disabled for this PVC)
>    * Existing, “Available”, PVs (that do not have a specified storageClassName) will be considered for binding to the PVC

The current behavior of Che is "dnyamic provisioning"
So we must not call `withStorageClassName()` if `che.infra.kubernetes.pvc.storage_class_name` is empty.

(My choice causes the another issue that no one can disable dynamic provisioning. But it isn't serious in usual use cases in Che.)

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java [hunk 5]
```diff
@@ -95,12 +96,26 @@ public static void putSelector(Service target, String key, String value) {
...
+      String name, String accessMode, String quantity, String storageClassName) {
+    SpecNested<PersistentVolumeClaimBuilder> specs =
+        new PersistentVolumeClaimBuilder()
+            .withNewMetadata()
+            .withName(name)
+            .endMetadata()
+            .withNewSpec()
+            .withAccessModes(accessMode);
+    if (storageClassName != null && !storageClassName.equals("")) {
```
[the remark this answers is not in the document]
Comment (monaka) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java]: It's better. I fixed.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategyTest.java [hunk 6]
```diff
@@ -113,6 +116,32 @@ public void shouldReturnPVCPerWorkspace() throws Exception {
...
```
Comment (monaka) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategyTest.java:119]: @amisevsk I added the test case of the default behavior here.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategyTest.java (hunk 2 above)
Comment (monaka) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategyTest.java:119]: I added the test case. (Please refer my comment with the mention to you)

**Comment by monaka** [comment]
ci-build

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 7]
```diff
@@ -104,7 +104,7 @@ che.workspace.default_memory_limit_mb=1024
 # RAM limit default for each sidecar that has no RAM settings in Che plugin configuration.
 che.workspace.sidecar.default_memory_limit_mb=128
 
-# Define image pulling strategy for sidecars.
+# Define image pulling strategy for sidecars./
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Do we really need this `/` here

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 8]
```diff
@@ -430,6 +430,10 @@ che.infra.kubernetes.pvc.precreate_subpaths=true
 # Defines the name of Persistent Volume Claim for che workspaces.
 che.infra.kubernetes.pvc.name=claim-che-workspace
 
+# Defines the storage class of Persistent Volume Claim for the workspaces.
+# Empty strings means "use default".
+che.infra.kubernetes.pvc.storage_class_name=
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:435]: @monaka Could you please contribute the corresponding changes to helm chart, to be able to configure `storage_class_name` with helm parameter? Thanks.

**Review by sleshchenko — COMMENTED** [review]
Thanks for contribution
LGTM :+1: 

It would be so nice to have helm chart updated to get your feature usable.

**Comment by sleshchenko** [comment]
ci-build

**Comment by sleshchenko** [comment]
ci-test

**Comment by sleshchenko** [comment]
@eclipse/eclipse-che-qa Could you please verify the test report? Thanks

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 7 above)
Comment (monaka) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Oops. No, I reverted it.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 8 above)
Comment (monaka) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:435]: I agree it's better.
At the same, I have some additional patches to Helm charts `deployments.yaml` and `configmap.yaml`.
So I want to send them (including storage_class_name) as the individual pull request.

**Review by sleshchenko — APPROVED** [review]
LGTM
Thanks for the contribution

**Comment by sleshchenko** [comment]
I'm going to merge this PR after QE team approval.
@metlos @amisevsk @garagatyi Could you please review it as well?

**Comment by sleshchenko** [comment]
ci-build

**Comment by sleshchenko** [comment]
ci-test

**Comment by sleshchenko** [comment]
Return tests after merging with actual master

**Review by garagatyi — APPROVED** [review]

**Review by amisevsk — APPROVED** [review]

**Comment by SkorikSergey** [comment]
ci-build

**Comment by SkorikSergey** [comment]
Selenium tests execution on Eclipse Che Multiuser on OCP (https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1640//Selenium_20tests_20report/) doesn't show any regression against this Pull Request.

**Comment by sleshchenko** [comment]
ci-build

**Pull request merged**